EIPS/eip-7495.md

```diff
@@ -0,0 +1,165 @@
+discussions-to: https://ethereum-magicians.org/t/eip-7495-ssz-partialcontainer/15476
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-08-18
+---
+
+## Abstract
+
+This EIP introduces a new [Simple Serialize (SSZ) type](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/ssz/simple-serialize.md) to represent `PartialContainer[T, N]` values.
+
+A `PartialContainer[T, N]` is an SSZ `Container` with stable serialization and merkleization even when individual fields become optional or new fields are introduced in the future.
+
+## Motivation
+
+Partial containers are currently not representable in SSZ. Adding support provides these benefits:
+
+1. **Stable signatures:** Signing roots derived from a `PartialContainer[T, N]` never change. In the context of Ethereum, this is useful for transaction signatures that are expected to remain valid even when future updates introduce additional transaction fields. Likewise, the overall transaction root remains stable and can be used as a perpetual transaction ID.
+
+2. **Stable merkle proofs:** Merkle proof verifiers that check specific fields of a `PartialContainer[T, N]` do not need continuous updating when future updates introduce additional fields. Common fields always merkleize at the same [generalized indices](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/ssz/merkle-proofs.md).
+
+3. **Compact serialization:** SSZ serialization is compact; inactive fields do not consume space.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### Type definition
+
+`PartialContainer[T, N]` defines a wrapper around an SSZ `Container` type indicated by `T`. `N` indicates the maximum number of fields to which `T` can grow in the future. `N` MUST be `> 0`.
+
+When wrapped in a `PartialContainer[T, N]`, `T` MAY define fields of type `Optional[E]`. Such fields can either represent a value of SSZ type `E`, or absence of a value (indicated by `None`). The [default value](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/ssz/simple-serialize.md#default-values) of an `Optional[E]` is `None`.
+
+For the purpose of serialization, `PartialContainer[T, N]` is always considered ["variable-size"](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/ssz/simple-serialize.md#variable-size-and-fixed-size) regardless of `T`.
+
+### Stability guarantees
+
+The serialization and merkleization of a `PartialContainer[T, N]` remains stable as long as:
+
+- The maximum capacity `N` does not change
+- The order of fields within `T` does not change
+- New fields are always added to the end of `T`
+- Required fields of `T` remain required `E`, or become an `Optional[E]`
```

Comment:
can this happen?: required => optional => required (and so on...) as this and below points imply infinite change

Reply:
Yes, a new fork could relax validity constraints, while a subsequent fork could require the field to be present again. Whether that makes sense is up to the application, but the serialization format proposed here will be stable across both directions (from optional to required and vice versa), essentially providing maximum flexibility when it comes to future design space.

In fact, an implementation could simply treat all fields as optional always, if they prefer to enforce constraints in the application layer instead of the serialization library. Another implementation may go the other direction and also push down an invariant callback to ensure that only valid combinations are accepted; e.g., a receipt can contain _either_ an intermediate state root, _or_ a status code, but not both at the same time. Yet another implementation may try to use union types to represent various valid combinations; point being is, none of these implementation details leak into the serialization representation.

Reply:
--> https://ethereum-magicians.org/t/eip-7495-ssz-partialcontainer/15476 is a better location for discussing technical points.